libcontainer/state_linux.go

```diff
@@ -36,12 +35,6 @@ type containerState interface {
 }
 
 func destroy(c *Container) error {
-	if !c.config.Namespaces.Contains(configs.NEWPID) ||
-		c.config.Namespaces.PathOf(configs.NEWPID) != "" {
-		if err := signalAllProcesses(c.cgroupManager, unix.SIGKILL); err != nil {
-			logrus.Warn(err)
-		}
-	}
```

Comment:
I'm not sure removing this entirely is the right move here.

In runc, `runc delete` handles the killing separately (and because of `--force` we don't want the killing to always be automatic -- but because `c.Destroy` has done this automatically we have checks in `runc delete` to avoid doing this automatically). However, for libcontainer users the current API (and IMHO the API which makes the most sense) is such that `c.Destroy()` will also kill the container processes as part of destroying the container. The alternative is AFAICS that `c.Destroy` either succeeds or errors out in a way that you cannot easily figure out that you forgot to call `c.Signal(unix.SIGKILL)`.

I think swapping this code for `c.Signal(unix.SISKILL)` would be most appropriate here, but I suspect it'd need to be put in `c.Destroy` instead because of the locking (presumably the reason this implemented a miniature version of the `sharedPidns` checks is because of the locking? I don't see why you'd want two different set of semantics about `sharedPidns` and I suspect this was done by accident.) Or alternatively, make `c.Signal` a locking wrapper around `c.signal` and then call `c.signal` here (this might be more preferable).

@littlelightlittlefire Do you guys call `c.Destroy()` without calling `c.Signal(unix.SIGKILL)` first?

Reply:
We go through and carefully kill everything first, wait for the initproc to die, before calling container destory.

Reply:
> we do not change the libcontainer API here

I was talking about the `libct: move killing logic to container.Signal` commit, specifically.

After this PR, though, the two changes to libcontainer API are:
 - `container.Signal` loses the `all bool` argument;
 - a user must implement their own child reaper.

Reply:
I did review this commit-by-commit (as I usually do), my issue is that calling `c.Destroy` will now no longer trigger a kill of the processes in the container.

To be fair, looking at it again I missed that `(*runningState).destroy()` fails if the container is in the `Running` state -- but at the moment `destroy(c)` is called if pid1 exists but the container is in the `Created` state. So while my concern about running containers wasn't correct (though I think that the API for `c.Destroy()` should also destroy running containers -- I don't understand why this logic is somewhat duplicated for `runc delete` as well), for `Created` containers this change will cause the pid1 process to not be killed on `c.Destroy` AFAICS.

In particular, consider that `destroy(c)` is called in `(*runningState).destroy()` which is called from `c.Destroy()` (and `c.Destroy()` doesn't do any process killing, it just calls `c.s.destroy()`).

Reply:
Ah, I see what you mean. Sorry, I misread part of your comment and didn't notice the other restrictions on `(*createdState).destroy()`. :sweat: